packages/xeus/src/worker.ts

```diff
@@ -173,37 +191,323 @@ export abstract class EmpackedXeusRemoteKernel extends XeusRemoteKernelBase {
+          specStatus.message = `Package ${pkgName} is not in the installed list`;
+        }
+        result.push(specStatus);
+      }
+    });
+    return result;
+  }
+
+  protected getInstalledPackagesMap() {
+    const newInstalledPackagesMap: { [name: string]: ISolvedPackage } = {};
+    for (const newInstalledPkg of Object.values(this._installedPackages)) {
+      newInstalledPackagesMap[newInstalledPkg.name] = newInstalledPkg;
+    }
+    return newInstalledPackagesMap;
+  }
+
+  protected updateCurrentSpecs(
+    specs: string[],
+    command: 'install' | 'uninstall'
+  ): {
+    specs: { [key: string]: string };
+    installed: ISolvedPackages;
+    status: number;
+  } {
+    let newSpecs = { ...this._currentSpecs };
+
+    /* we need to collect requested pip specs to know when we need to delete all packages
+       if a user tried to install and after they decided to delete pip packages
+       and if specs array is empty and pip specs array is not, then we should not clean an enviroment
+    */
+    let updatedInstalled: ISolvedPackages = { ...this._installedPackages };
+    const newInstalledPackagesMap: { [name: string]: ISolvedPackage } =
+      this.getInstalledPackagesMap();
+
+    let status: number = 1;
+
+    switch (command) {
+      case 'install':
+        newSpecs = this.addSpecs(specs, newSpecs);
+        break;
+      case 'uninstall': {
+        if (specs.length) {
+          const data = this.deleteSpecs(
+            specs,
+            newSpecs,
+            newInstalledPackagesMap,
+            updatedInstalled
+          );
+          newSpecs = data.specs;
+          updatedInstalled = data.installed;
+        } else {
+          status = 0;
+        }
+
+        break;
+      }
+      default:
+        break;
+    }
+
+    return {
+      specs: newSpecs,
+      installed: updatedInstalled,
+      status
+    };
+  }
+
   protected async install(
     channels: string[],
     specs: string[],
     pipSpecs: string[]
   ) {
-    if (specs.length || pipSpecs.length) {
-      try {
-        const newPackages = await solve({
-          ymlOrSpecs: specs,
-          installedPackages: this._installedPackages,
-          pipSpecs,
-          channels,
-          logger: this.logger
-        });
-
-        await this._reloadPackagesInFS({
-          ...newPackages.condaPackages,
-          ...newPackages.pipPackages
-        });
-      } catch (error: any) {
-        this.logger?.error(error.stack);
+    let data: {
+      specs: { [key: string]: string };
+      installed: ISolvedPackages;
+      status: number;
+    } = {
+      specs: {},
+      installed: this._installedPackages,
+      status: 1
+    };
+    if (specs.length) {
+      data = this.updateCurrentSpecs(specs, 'install');
+    }
+
+    try {
+      const newPackages = await solve({
+        ymlOrSpecs: Object.values(data.specs),
+        installedPackages: data.installed,
+        pipSpecs,
+        channels,
+        logger: this.logger
+      });
```

Comment:
We need to do something like the following. Otherwise we'll run the rattler solver with doing `%pip install bla`, which is not something we want:


```suggestion
      const command = specs?.length ? 'conda' : 'pip';
      let newPackages: {
        condaPackages: ISolvedPackages;
        pipPackages: ISolvedPackages;
      };

      switch (command) {
        case 'conda': {
          newPackages = await solve({
            ymlOrSpecs: Object.values(data.specs),
            installedPackages,
            channels: newChannels,
            logger
          });
          break;
        }
        case 'pip': {
          newPackages = await solve({
            installedPackages,
            pipSpecs,
            logger
          });
          break;
        }
      }
```